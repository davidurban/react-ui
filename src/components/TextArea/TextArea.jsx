import PropTypes from 'prop-types';
import React from 'react';
import UserInputBase from '../UserInputBase';

const TextArea = (props) => {
  const {
    placeholder,
    rows,
    ...higherOrderProps
  } = props;

  return (
    <textarea
      {...higherOrderProps}
      placeholder={placeholder}
      rows={rows}
    />
  );
};

TextArea.defaultProps = {
  placeholder: null,
  rows: 3,
};

TextArea.propTypes = {
  placeholder: PropTypes.string,
  rows: PropTypes.number,
};

export default UserInputBase(TextArea);
