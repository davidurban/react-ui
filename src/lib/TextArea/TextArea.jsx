import PropTypes from 'prop-types';
import React from 'react';
import styles from './TextArea.scss';

const TextArea = (props) => {
  let labelClass = styles.label;
  if (props.isLabelVisible) {
    if (props.required) {
      labelClass = styles.isLabelRequired;
    }
  } else {
    labelClass = styles.isLabelHidden;
  }

  return (
    <div className={styles.root}>
      <label htmlFor={props.fieldId}>
        <div className={labelClass}>
          {props.label}
        </div>
        <textarea
          id={props.fieldId}
          placeholder={props.placeholder}
          required={props.required}
          rows={props.rows}
          value={props.value}
          onChange={props.changeHandler}
          className={props.error ? styles.isInputInvalid : styles.input}
        />
      </label>
      {props.description && (
        <div className={styles.description}>
          {props.description}
        </div>
      )}
      {props.error && (
        <div className={styles.error}>
          {props.error}
        </div>
      )}
    </div>
  );
};

TextArea.defaultProps = {
  changeHandler: null,
  description: null,
  error: null,
  isLabelVisible: true,
  placeholder: null,
  required: false,
  rows: 3,
  value: undefined,
};

TextArea.propTypes = {
  changeHandler: PropTypes.func,
  description: PropTypes.string,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  rows: PropTypes.number,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default TextArea;
