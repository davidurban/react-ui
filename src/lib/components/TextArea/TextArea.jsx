import PropTypes from 'prop-types';
import React from 'react';
import styles from './TextArea.scss';

const TextArea = (props) => {
  let labelVisibilityClass = '';
  let rootFullWidthClass = '';
  let rootLayoutClass = '';
  let rootRequiredClass = '';
  let rootSizeClass = '';
  let rootValidationStateClass = '';
  let rootVariantClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.fullWidth) {
    rootFullWidthClass = styles.isRootFullWidth;
  }

  if (props.layout === 'horizontal') {
    rootLayoutClass = styles.rootLayoutHorizontal;
  } else if (props.layout === 'vertical') {
    rootLayoutClass = styles.rootLayoutVertical;
  }

  if (props.required) {
    rootRequiredClass = styles.isRootRequired;
  }

  if (props.size === 'small') {
    rootSizeClass = styles.rootSizeSmall;
  } else if (props.size === 'medium') {
    rootSizeClass = styles.rootSizeMedium;
  } else if (props.size === 'large') {
    rootSizeClass = styles.rootSizeLarge;
  }

  if (props.validationState === 'invalid') {
    rootValidationStateClass = styles.isRootStateInvalid;
  } else if (props.validationState === 'valid') {
    rootValidationStateClass = styles.isRootStateValid;
  } else if (props.validationState === 'warning') {
    rootValidationStateClass = styles.isRootStateWarning;
  }

  if (props.variant === 'filled') {
    rootVariantClass = styles.rootVariantFilled;
  } else if (props.variant === 'outline') {
    rootVariantClass = styles.rootVariantOutline;
  }

  return (
    <div
      className={(`
        ${styles.root}
        ${rootFullWidthClass}
        ${rootLayoutClass}
        ${rootRequiredClass}
        ${rootSizeClass}
        ${rootValidationStateClass}
        ${rootVariantClass}
      `).trim()}
    >
      <label className={styles.container} htmlFor={props.fieldId}>
        <div
          className={(`
            ${styles.label}
            ${labelVisibilityClass}
          `).trim()}
        >
          {props.label}
        </div>
        <div className={styles.inputContainer}>
          <textarea
            className={styles.input}
            cols={props.cols}
            disabled={props.disabled}
            id={props.fieldId}
            onChange={props.changeHandler}
            placeholder={props.placeholder}
            required={props.required}
            rows={props.rows}
            value={props.value}
          />
          {props.variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
      </label>
      {props.helperText && (
        <div className={styles.helperText}>
          {props.helperText}
        </div>
      )}
    </div>
  );
};

TextArea.defaultProps = {
  changeHandler: null,
  cols: null,
  disabled: false,
  fullWidth: false,
  helperText: null,
  isLabelVisible: true,
  layout: 'vertical',
  placeholder: null,
  required: false,
  rows: 3,
  size: 'medium',
  validationState: null,
  value: undefined,
  variant: 'outline',
};

TextArea.propTypes = {
  changeHandler: PropTypes.func,
  cols: PropTypes.number,
  disabled: PropTypes.bool,
  fieldId: PropTypes.string.isRequired,
  fullWidth: PropTypes.bool,
  helperText: PropTypes.string,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  rows: PropTypes.number,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  value: PropTypes.string,
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export default TextArea;
