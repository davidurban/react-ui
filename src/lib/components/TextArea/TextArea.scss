@import '../../Theme/mixins';
@import '../../Theme/variables';
@import '../../Theme/root';

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  padding: 0 0.25rem 0 0;
  margin-bottom: 1rem;
}

.label {
  display: inline-block;
  margin-bottom: 0.25rem;
}

.isLabelRequired {
  composes: label;

  &::after {
    content: ' *';
  }
}

.input {
  display: block;
  box-sizing: border-box;
  width: 100%;
  padding: 0.25rem;
  font-family: var(--font-family);
  color: $rui-textarea-text-color;
  border: $rui-border-width solid $rui-textarea-border-color;
  border-radius: $rui-border-radius;
  background: $rui-textarea-bg-color;
  resize: vertical;

  &:focus {
    @include rui-include-outline-style-form-fields();
  }

  &:disabled {
    @include rui-disabled-form-fields();
  }
}

.isInputInvalid {
  composes: input;
  border-color: $rui-textarea-invalid-border-color;
}

.description,
.error {
  font-size: 0.8rem;
}

.error {
  min-height: 1rem;
  color: $rui-input-invalid-color;
}
