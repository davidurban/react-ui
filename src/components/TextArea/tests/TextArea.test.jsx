import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import TextArea from '../index';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<TextArea fieldId="test" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<TextArea
      label="label"
      disabled
      fieldId="test"
      value="value"
      helpText="some help"
      errors={['some error', 'another error']}
      placeholder="placeholder"
      rows={5}
      required
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onChange()', () => {
    const spy = sinon.spy();
    const component = mount(<TextArea
      fieldId="test"
      onChange={spy}
    />);

    component
      .find('textarea')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'test' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
