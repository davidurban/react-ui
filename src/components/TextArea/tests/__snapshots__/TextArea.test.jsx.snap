// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div>
  <TextArea
    disabled={false}
    id="test"
    onChange={null}
    placeholder={null}
    required={false}
    rows={3}
  />
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div>
  <label
    htmlFor="test"
  >
    label
  </label>
  <TextArea
    disabled={true}
    id="test"
    onChange={null}
    placeholder="placeholder"
    required={true}
    rows={5}
    value="value"
  />
  <div>
    some help
  </div>
  <div
    key="some error"
  >
    some error
  </div>
  <div
    key="another error"
  >
    another error
  </div>
</div>
`;
